pydantic_extra_types/mac_address.py

```diff
@@ -0,0 +1,111 @@
+# Mac Address Parsing and Validation
+# This code provides functionality to parse and validate Mac Addresses in different formats,
+# such as IEEE 802 MAC-48, EUI-48, EUI-64, or a 20-octet format. It includes a `MacAddress` class
+# that represents a Mac Address and provides methods for conversion, validation, and serialization.
+# The code also includes a `validate_mac_address` function that takes a byte value representing a Mac Address
+# and returns the parsed Mac Address as a string.
+# Various custom error classes, such as `PydanticCustomError`, are used to handle specific validation errors.
+
+# The code is organized into the following components:
+
+# - MacAddress: A class representing a Mac Address. It provides methods for initialization, serialization,
+#   validation, and conversion.
+
+# - validate_mac_address: A function that takes a byte value representing a Mac Address.
+#   It supports parsing Mac Addresses in different formats based on specific rules and validations.
+
+from __future__ import annotations
+
+from typing import Any, Callable, Union
+
+from pydantic_core import CoreSchema, PydanticCustomError, core_schema
+
+MacAddressType = Union[str, bytes]
+
+
+class MacAddress(str):
+    """
+    Represents a mac address - IEEE 802 MAC-48, EUI-48, EUI-64, or a 20-octet
+    """
+
+    __slots__ = '_mac_address'
+
+    def __init__(self, value: MacAddressType) -> None:
+        self._mac_address: str
+        if isinstance(value, bytes):
+            self._mac_address = validate_mac_address(value)
+        elif isinstance(value, str):
+            self._mac_address = validate_mac_address(value.encode())
+        else:
+            raise PydanticCustomError(
+                'mac_address_error',
+                'value is not a valid Mac Address: value must be a string or bytes',
+            )
+
+    @classmethod
+    def __get_pydantic_core_schema__(
+        cls, source: type[Any], handler: Callable[[Any], CoreSchema]
+    ) -> core_schema.CoreSchema:
+        return core_schema.general_plain_validator_function(cls._validate)
+
+    @classmethod
+    def _validate(cls, __input_value: Any, _: Any) -> MacAddress:
+        return cls(__input_value)
+
+
+def validate_mac_address(value: bytes) -> str:
+    """
+    Validate a Mac Address from the provided byte value.
+
+    Args:
+        value (bytes): The byte value representing the Mac Address.
+
+    Returns:
+        str: The parsed Mac Address as a string.
+
+    Raises:
+        PydanticCustomError: If the value is not a valid Mac Address.
+    """
+    if len(value) < 14:
+        raise PydanticCustomError(
+            'mac_address_len',
+            'Length for a {mac_address} mac address must be {required_length}',
+            {'mac_address': value.decode(), 'required_length': 14},
+        )
+
+    if value[2] == ord(':') or value[2] == ord('-'):
```

Comment:
``` suggestion
    if value[2] in [ord(':'), ord('-')]:
```

It's better to Replace multiple comparisons of the same variable with the `in` operator